Feat/adapt share project third step

positiveimpact opened the pull request:
This PR finished up the necessary changes that need to be made to the "Share project" code to enable the event calendar (see #1189). 
It includes the following changes:
- Enter project name in step 3 of share project
- Use new Date pickers from upgraded Material ui - show different date options dependent on type. Added manual date validation because options like `required` aren't supported by the new package
- Allow putting exact locations (e.g. townhall San Francisco). Before it was only possible to specify the city
- Allow putting additional info to locations (e.g. which room in a building an event is in)

## To test:
- Run migrations
- Run `create_project_type_test_data` section of the `create_test_data` management script
- Go through the "Share project" flow once for each project type, see the different options
- Share one project of each type

## Icons used for the three project types to use in your local db
Project type icon
![project_type_icon](https://github.com/climateconnect/climateconnect/assets/10531079/6f1b16fe-19c1-42ed-9838-cab7857721e8)

Idea type icon
![idea_type_icon](https://github.com/climateconnect/climateconnect/assets/10531079/0c299fb1-45cf-4bb0-b722-2fed45ad88b3)

Event type icon
![event_type_icon](https://github.com/climateconnect/climateconnect/assets/10531079/f750958c-4d73-4da1-b53b-a4c5f88b2e82)

ddhanesha commented:
@positiveimpact running migrations are failing. getting the following error.

```python
Operations to perform:
  Apply all migrations: admin, auth, chat_messages, climate_match, climateconnect_api, contenttypes, django_celery_beat, hubs, ideas, knox, location, organization, sessions
Running migrations:
  Applying location.0010_auto_20230704_1837... OK
  Applying location.0011_alter_location_exact_address... OK
  Applying organization.0102_auto_20230327_0613...Traceback (most recent call last):
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
psycopg2.errors.UndefinedColumn: column organization_projecttypes.help_text_de_translation does not exist
LINE 1: ...lation", "organization_projecttypes"."help_text", "organizat...
                                                             ^


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ddhanesha/repos/local_cc/climateconnect/backend/./manage.py", line 23, in <module>
    main()
  File "/Users/ddhanesha/repos/local_cc/climateconnect/backend/./manage.py", line 19, in main
    execute_from_command_line(sys.argv)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/core/management/base.py", line 89, in wrapped
    res = handle_func(*args, **kwargs)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/core/management/commands/migrate.py", line 244, in handle
    post_migrate_state = executor.migrate(
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/migrations/executor.py", line 227, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/migrations/migration.py", line 126, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/migrations/operations/fields.py", line 104, in database_forwards
    schema_editor.add_field(
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/base/schema.py", line 490, in add_field
    definition, params = self.column_sql(model, field, include_default=True)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/base/schema.py", line 237, in column_sql
    default_value = self.effective_default(field)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/base/schema.py", line 324, in effective_default
    return field.get_db_prep_save(self._effective_default(field), self.connection)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/base/schema.py", line 303, in _effective_default
    default = field.get_default()
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/models/fields/related.py", line 960, in get_default
    field_default = super().get_default()
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/models/fields/__init__.py", line 850, in get_default
    return self._get_default()
  File "/Users/ddhanesha/repos/local_cc/climateconnect/backend/organization/models/status.py", line 151, in get_default_pk
    project_type, created = cls.objects.get_or_create(
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/models/manager.py", line 85, in manager_method
    return getattr(self.get_queryset(), name)(*args, **kwargs)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/models/query.py", line 581, in get_or_create
    return self.get(**kwargs), False
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/models/query.py", line 431, in get
    num = len(clone)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/models/query.py", line 262, in __len__
    self._fetch_all()
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/models/query.py", line 1324, in _fetch_all
    self._result_cache = list(self._iterable_class(self))
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/models/query.py", line 51, in __iter__
    results = compiler.execute_sql(chunked_fetch=self.chunked_fetch, chunk_size=self.chunk_size)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/models/sql/compiler.py", line 1175, in execute_sql
    cursor.execute(sql, params)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 98, in execute
    return super().execute(sql, params)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/sentry_sdk/integrations/django/__init__.py", line 562, in execute
    return real_execute(self, sql, params)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.ProgrammingError: column organization_projecttypes.help_text_de_translation does not exist
LINE 1: ...lation", "organization_projecttypes"."help_text", "organizat...
                                                             ^
```

ddhanesha commented on backend/organization/models/status.py:
```diff
@@ -70,3 +70,86 @@ class Meta:
+
+
+class ProjectTypes(models.Model):
+    name = models.CharField(
```
Can users define custom project types? If yes, I would like to understand it better. But if its maintained by us then we do not need a database table. I think we can use constants or enums (https://docs.python.org/3/library/enum.html) You can store the icons as static files as well.

positiveimpact replied:
Currently they don't and you're right we should use an enum for this. I was thinking this was more dynamic and easier to make changes for non-devs but these advantages don't outweight the disadvantage of using a database for static data at all. I guess I just went with the same concept we used for the old "Project Status". I'll make the necessary changes in the backend, thanks for pointing this out.

Would you still keep endpoints for this and just make them access the enum in the backend or additionally store the enum in the frontend? :thinking:

ddhanesha commented on backend/climateconnect_main/utility/general.py:
```diff
@@ -55,3 +55,12 @@ def get_allowed_hosts(ALLOWED_HOSTS_ENV):
     return allowed_hosts
+
+
+def get_prop_in_language(obj, prop_name: str, language_code: str) -> str:
```
Can you add `obj: Queryset` as a type?

positiveimpact commented:
Fixed migrations :+1: @ddhanesha Thanks for pointing out that they are failing.

positiveimpact commented on backend/location/models.py:
```diff
@@ -10,6 +10,22 @@ class Location(models.Model):
+        max_length=1024,
+        blank=True,
+        null=True,
+    )
```
This does introduce street address. Tbh I thought the fields exact_address and place_name were sufficient looking at many examples of address data from Nominatim. Since we store place_id and osm_id we can always catch the full data from Nominatim again, I don't think we need more in our database than we use.
What would be the advantage of storing exact locations in a different table? They have the same data structure and seperate place_id's and osm_id's and geodata so they are always unique. They just additionally fill one or both of the newly added fields. Adding different tables would make location such much more complicated because of course if somebody types in "San Francisco" you also want to show events taking place in "San Francisco city hall". But if you had those locations in two different tables you'd have to search both tables to find all relevant locations.

ddhanesha replied:
Okay, then it makes sense to keep in the same table.

positiveimpact commented:
Addressed all comments and converted ProjectTypes to an enum :+1: The way I programmed it it's still being retrieved through the API by the frontend however. @ddhanesha

ddhanesha commented:
> Fixed migrations 👍 @ddhanesha Thanks for pointing out that they are failing.

Hey I am still getting the migration error.
```python
psycopg2.errors.NotNullViolation: column "project_type_id" of relation "organization_project" contains null values


The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/Users/ddhanesha/repos/local_cc/climateconnect/backend/./manage.py", line 23, in <module>
    main()
  File "/Users/ddhanesha/repos/local_cc/climateconnect/backend/./manage.py", line 19, in main
    execute_from_command_line(sys.argv)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/core/management/__init__.py", line 419, in execute_from_command_line
    utility.execute()
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/core/management/__init__.py", line 413, in execute
    self.fetch_command(subcommand).run_from_argv(self.argv)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/core/management/base.py", line 354, in run_from_argv
    self.execute(*args, **cmd_options)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/core/management/base.py", line 398, in execute
    output = self.handle(*args, **options)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/core/management/base.py", line 89, in wrapped
    res = handle_func(*args, **kwargs)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/core/management/commands/migrate.py", line 244, in handle
    post_migrate_state = executor.migrate(
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/migrations/executor.py", line 117, in migrate
    state = self._migrate_all_forwards(state, plan, full_plan, fake=fake, fake_initial=fake_initial)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/migrations/executor.py", line 147, in _migrate_all_forwards
    state = self.apply_migration(state, migration, fake=fake, fake_initial=fake_initial)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/migrations/executor.py", line 227, in apply_migration
    state = migration.apply(state, schema_editor)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/migrations/migration.py", line 126, in apply
    operation.database_forwards(self.app_label, schema_editor, old_state, project_state)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/migrations/operations/fields.py", line 104, in database_forwards
    schema_editor.add_field(
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/base/schema.py", line 522, in add_field
    self.execute(sql, params)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/base/schema.py", line 145, in execute
    cursor.execute(sql, params)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 98, in execute
    return super().execute(sql, params)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/sentry_sdk/integrations/django/__init__.py", line 562, in execute
    return real_execute(self, sql, params)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 66, in execute
    return self._execute_with_wrappers(sql, params, many=False, executor=self._execute)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 75, in _execute_with_wrappers
    return executor(sql, params, many, context)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/utils.py", line 90, in __exit__
    raise dj_exc_value.with_traceback(traceback) from exc_value
  File "/Users/ddhanesha/repos/local_cc/cc_env/lib/python3.9/site-packages/django/db/backends/utils.py", line 84, in _execute
    return self.cursor.execute(sql, params)
django.db.utils.IntegrityError: column "project_type_id" of relation "organization_project" contains null values
```

ddhanesha reviewed (approved):
@positiveimpact Approved! 

Hey, the other PR needs to be updated because this PR has mirgration fix
